pkg/authentication/context.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+)
+
+type principalContextKey int
+
+const stateSize = 18
+
+var (
+	letterRunes         = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	PrincipalContextKey principalContextKey
+	otelHTTPClient      = http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}
+)
+
+type Principal struct {
+	Subject string `json:"sub"`
```

Comment:
Question out of scope for this PR: Shouldn't we use the user's email instead of the `sub`? This would be needed for the exposing the user's ID in the UI